File: domain/src/main/java/com/chronos/mobile/domain/SecureCredentialStore.kt
package com.chronos.mobile.domain

import com.chronos.mobile.domain.model.AuthSnapshot
import com.chronos.mobile.domain.model.SavedCredentialState
import com.chronos.mobile.domain.result.AppResult
import javax.crypto.Cipher
import kotlinx.coroutines.flow.Flow

interface SecureCredentialStore {
    val savedCredentialState: Flow<SavedCredentialState>

    fun isReusableCredentialProtectionAvailable(): Boolean

    fun createSaveCipher(): AppResult<Cipher>

    suspend fun saveCredential(
        account: String,
        password: String,
        cipher: Cipher,
    ): AppResult<Unit>

    fun createUnlockCipher(): AppResult<Cipher>

    suspend fun unlock(cipher: Cipher): AppResult<AuthSnapshot>

    suspend fun clearCredential()
}

File: src/lib/domain/interfaces/secure-credential-store.ts
import type { SavedCredentialState } from '$lib/models/auth';
import type { AppResult } from '../result/app-result';

export interface SecureCredentialStore {
	subscribeSavedCredentialState(listener: (state: SavedCredentialState) => void): () => void;
	saveCredential(account: string, password: string, unlockToken: string): Promise<AppResult<void>>;
	unlockCredential(unlockToken: string): Promise<AppResult<{ account: string; password: string }>>;
	clearCredential(): Promise<AppResult<void>>;
	prepareSave(): Promise<AppResult<string>>;
	prepareUnlock(): Promise<AppResult<string>>;
}
